core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -0,0 +1,280 @@
+  }
+
+  private[server] def validateAndConvertForPersistence(brokerId: Option[Int], props: Properties): Properties = CoreUtils.inReadLock(lock) {
+    validateConfigTypes(brokerId, props, logError = false)
+    val newProps = mutable.Map[String, String]()
+    newProps ++= staticBrokerConfigs
+    brokerId match {
+      case Some(_) =>
+        overrideProps(newProps, dynamicDefaultConfigs)
+        overrideProps(newProps, props.asScala)
+      case None =>
+        overrideProps(newProps, props.asScala)
+        overrideProps(newProps, dynamicBrokerConfigs)
+    }
+    processConfig(newProps, validateOnly = true)
+    toPersistentProps(props)
+  }
+
+  private def validateConfigTypes(brokerId: Option[Int], props: Properties, logError: Boolean): Unit = {
+    try {
+      val securityConfigsWithoutPrefix = dynamicSecurityConfigs.filter(props.containsKey)
+      if (securityConfigsWithoutPrefix.nonEmpty)
+          throw new ConfigException(s"Invalid configs $securityConfigsWithoutPrefix: security configs can be dynamically updated only using listener prefix")
+      val baseProps = new Properties
+      props.asScala.foreach {
+        case (listenerConfigRegex(baseName), v) => baseProps.put(baseName, v)
+        case (k, v) => baseProps.put(k, v)
+      }
+      DynamicConfig.Broker.validate(baseProps)
+    } catch {
+      case e: Exception =>
+        if (logError)
+          error(s"Dynamic default broker config is invalid: $props, these configs will be ignored", e)
+        throw e
+    }
+  }
+
+  private def configToMap(config: Map[String, _]): immutable.Map[String, String] = {
+    config.filter(_._2 != null).map {
+      case (k, v: String) => (k, v)
+      case (k, v: Password) => (k, v.value)
+      case (k, v: java.util.List[_]) => (k, ConfigDef.convertToString(v, ConfigDef.Type.LIST))
+      case (k, v: Class[_]) => (k, ConfigDef.convertToString(v, ConfigDef.Type.CLASS))
+      case (k, v) => (k, String.valueOf(v))
+    }.toMap
+  }
+
+  private def updatedConfigs(newProps: java.util.Map[String, _], currentProps: java.util.Map[_, _]): mutable.Map[String, _] = {
+    newProps.asScala.filter {
+      case (k, v) => v != currentProps.get(k)
+    }
+  }
+
+  private def overrideProps(props: mutable.Map[String, String], propsOverride: mutable.Map[String, String]): Unit = {
+    propsOverride.foreach { case (k, v) =>
+      brokerConfigSynonyms(k, matchListenerOverride = false).foreach(props.remove)
+      props.put(k, v)
+    }
+  }
+
+  private def updateCurrentConfig(): Unit = {
+    val newProps = mutable.Map[String, String]()
+    newProps ++= staticBrokerConfigs
+    overrideProps(newProps, dynamicDefaultConfigs)
+    overrideProps(newProps, dynamicBrokerConfigs)
+    currentConfig = processConfig(newProps, validateOnly = false)
+  }
+
+  private def processConfig(newProps: mutable.Map[String, String], validateOnly: Boolean): KafkaConfig = {
+    val newConfig = new KafkaConfig(newProps.asJava, true, Some(this))
```

Comment:
Do we want logging enabled if we are only validating? We skip error logging below if `validateOnly` is enabled.